### PR TITLE
C#: support serialization to IBufferWriter

### PR DESCRIPTION
**Pull request by jtattermusch**
Allow writing to SerializationContext through `IBufferWriter<byte>` interface.
- allows writing directly to a native slice buffer, eliminating overhead of copying chunks of memory between `byte[]` and `grpc_byte_buffer`
- allows slice-by-slice serialization of bigger payloads
- current way of serializing data requires user to allocate a new byte array each time they serialize a message, which can cause a lot of GC pressure especially if the payload is big. The new API allows to overcome this limitation
- grpc-dotnet implementation should be able to provide access to the http response in form of IBufferWriter as well.

Work remaining:
- make sure the default protobuf serialization uses the more efficient API when possible.

### CODE REVIEW

**Comment by jtattermusch** [comment]
@apolcyn  this is ready for review.
@JunTaoLuo  @JamesNK  feel free to review.

File: src/csharp/Grpc.Core/Internal/AsyncCallBase.cs [hunk 1]
```diff
@@ -115,23 +115,25 @@ protected void InitializeInternal(INativeCall call)
...
-                }
+                    GrpcPreconditions.CheckState(started);
+                    var earlyResult = CheckSendAllowedOrEarlyResult();
+                    if (earlyResult != null)
+                    {
+                        return earlyResult;
+                    }
 
-                call.StartSendMessage(SendCompletionCallback, payload, writeFlags, !initialMetadataSent);
+                    call.StartSendMessage(SendCompletionCallback, payload, writeFlags, !initialMetadataSent);
```
[the remark this answers is not in the document]
Comment (jtattermusch) [reply src/csharp/Grpc.Core/Internal/AsyncCallBase.cs]: I think the indentation is fine, it's just the git diff is rendering this being nested under a new `using` block weird.

**Comment by jtattermusch** [comment]
@apolcyn  addressed the comments.

@JamesNK  @JunTaoLuo  can you please review the additions to the SerializationContext API and confirm that this is going to work well with grpc-dotnet?

File: src/csharp/Grpc.Core.Api/SerializationContext.cs [hunk 2]
```diff
@@ -27,13 +28,31 @@ public abstract class SerializationContext
...
+        /// Gets buffer writer that can be used to write the serialized data. Once serialization is finished,
+        /// <c>Complete()</c> needs to be called.
+        /// </summary>
+        public virtual IBufferWriter<byte> GetBufferWriter()
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Complete the payload written to the buffer writer. Complete() can only be called once.
```
Comment (JamesNK) [inline src/csharp/Grpc.Core.Api/SerializationContext.cs]: nit for consistency with the doc above:
```suggestion
        /// Complete the payload written to the buffer writer. <c>Complete()</c> can only be called once.
```

**Comment by jtattermusch** [comment]
> For grpc-dotnet's perspective, we only really care about the two new virtual methods on SerializationContext. Is that correct? Looks ok to me.

Yes, I wanted to know if you're happy with the newly proposed methods and their behavior.
(and if the Complete() semantics makes sense to you).
> 
> I guess there isn't much way to performance test without a Protobuf serializer that supports IBufferWriter.

**Pull request merged**